psydac/linalg/solvers.py

```diff
@@ -1725,3 +1728,261 @@ def solve(self, b, out=None):
+        verbose = options["verbose"]
+        
+
+        assert isinstance(b, Vector)
+        assert b.space is domain
+
+        # First guess of solution
+        if out is not None:
+            assert isinstance(out, Vector)
+            assert out.space is codomain
+
+        x = x0.copy(out=out)
+        
+        # Extract local storage
+        r = self._tmps["r"]
+        v = self._tmps["v"]
+
+        # Internal objects of GMRES
+        H = self._H
+        Q = []
+        beta = []
+        sn = []
+        cn = []
+
+        # First values
+        A.dot( x , out=r)
+        r -= b
+
+        am = r.dot(r).real ** 0.5
+        beta.append(am)
+        Q.append(- r / am)       
+
+        if verbose:
+            print( "GMRES solver:" )
+            print( "+---------+---------------------+")
+            print( "+ Iter. # | L2-norm of residual |")
+            print( "+---------+---------------------+")
+            template = "| {:7d} | {:19.2e} |"
+            print( template.format( 1, am ) )
+
+        # Iterate to convergence
+        for k in range(maxiter):
+            if am < tol:
+                break
+
+            # run Arnoldi
+            self.arnoldi(k, Q)
+
+            # make the last diagonal entry in H equal to 0, so that H becomes upper triangular
+            self.apply_givens_rotation(k, sn, cn)
+
+            # update the residual vector
+            beta.append(- sn[k] * beta[k])
+            beta[k] *= cn[k]
+
+            am = abs(beta[k+1])
+            if verbose:
+                print( template.format( k+2, am ) )
+
+        if verbose:
+            print( "+---------+---------------------+")        
+        # calculate result
+        y = self.solve_triangular(H[:k, :k], beta[:k]) # system of upper triangular matrix
+
+        for i in range(k):
+            Q[i].copy(out=v)
+            v *= y[i]
+            x += v
+
+        # Convergence information
+        self._info = {'niter': k+1, 'success': am < tol, 'res_norm': am }
+        
+        return x
+    
+    def solve_triangular(self, T, d):
+        # Backwards substitution. Assumes T is upper triangular
+        k = T.shape[0]
+        y = np.zeros((k,), dtype=self._A.dtype)
+
+        for k1 in range(k):
+            temp = 0.
+            for k2 in range(1, k1 + 1):
+                temp += T[k - 1 - k1, k - 1 - k1 + k2] * y[k - 1 - k1 + k2]
+            y[k - 1 - k1] = ( d[k - 1 - k1] - temp ) / T[k - 1 - k1, k - 1 - k1]
+        
+        return y
+
+    def arnoldi(self, k, Q):
+        h = self._H[:k+2, k]
+
+        p = self._tmps["p"]
+        self._A.dot( Q[k] , out=p) # Krylov vector
+
+        lv = self._tmps["lv"]
+
+        for i in range(k + 1): # Modified Gram-Schmidt, keeping Hessenberg matrix
+            h[i] = p.dot(Q[i])
+            Q[i].copy(out=lv)
+            lv *= h[i]
+            p -= lv
+        
+        h[k+1] = p.dot(p).real ** 0.5
+        p /= h[k+1] # Normalize vector
+
+        Q.append(p.copy())
+
```

Comment:
This line creates lots of "temporaries", but from what I understood during our Psydac dev meeting last week there`s no way of avoiding this, right?

Reply:
There are two ways of avoiding temporaries (assuming the `solve` method is called multiple times):
1) Add the integer `restart` parameter and pre-allocate `restart` vectors (this would need a separate PR);
2) Reuse the vectors in `Q` between successive calls to `solve`:
 ```python
if len(Q) > k:
    p.copy(out = Q[k+1])
else:
    Q.append(p.copy())
```

Reply:
For the moment I take the second option